tt_eager/tt_dnn/op_library/moreh_matmul/moreh_matmul_op.cpp

```diff
@@ -115,30 +85,83 @@ inline Shape compute_output_shape(
+    if(!output_tensors.empty() && output_tensors.at(0).has_value()){
+        return {output_tensors.at(0).value()};
+    }
+    const auto& output_shapes = this->compute_output_shapes(input_tensors);
+    const auto& output_shape = output_shapes.at(0);
+
+    // Instead of ....
+    //     const Tensor& input_tensor = input_tensors.at(0);
+    //     const Tensor& other_tensor = input_tensors.at(0);
+    //     return {create_output_tensor(input_tensor, output_shape, this->output_mem_config)};
+    // could just do the following...
+    return {operation::generic_create_output_tensors(*this, input_tensors, input_tensors.at(0).dtype(), Layout::TILE, this->output_mem_config)};
+}
+
+void MorehMatmul::validate_with_output_tensors(const std::vector<Tensor> &input_tensors, const std::vector<std::optional<Tensor>>& output_tensors) const {
+    const auto& input_tensor = input_tensors.at(0);
+    const auto& other_tensor = input_tensors.at(1);
+    TT_ASSERT(
+        (input_tensor.layout() == Layout::TILE && other_tensor.layout() == Layout::TILE),
+        "Inputs to matmul must be tilized");
+
+    TT_ASSERT(
+        input_tensor.dtype() == DataType::BFLOAT16 || input_tensor.dtype() == DataType::BFLOAT8_B,
+        "Unsupported data format");
+    TT_ASSERT(
+        input_tensor.storage_type() == StorageType::DEVICE and other_tensor.storage_type() == StorageType::DEVICE,
+        "Operands to matmul need to be on device!");
+    TT_ASSERT(input_tensor.device() == other_tensor.device(), "Operands to matmul need to be on the same device!");
+    TT_ASSERT(
+        input_tensor.buffer() != nullptr and other_tensor.buffer() != nullptr,
+        "Operands to matmul need to be allocated in buffers on device!");
+
+
+    if(output_tensors.empty() || !output_tensors.at(0).has_value()){
+        // If the user decided to not use any optional output tensors, then this would be empty or would be a nullptr.
+        return;
+    }
+    const auto& input_shape = input_tensor.shape();
+    const auto& other_shape = other_tensor.shape();
+    const auto output_shape_required = compute_output_shape(input_shape, other_shape, this->transpose_input, this->transpose_other);
+    const auto& actual_shape = output_tensors.at(0).value().shape();
+    bool shape_ok = output_shape_required.rank() == actual_shape.rank();
+    for(size_t i=0; i < std::min(actual_shape.rank(),output_shape_required.rank()) ; i++){
+        shape_ok &= output_shape_required[i] <= actual_shape[i];
+    }
+    TT_ASSERT(shape_ok, fmt::format("The input tensors need a shape of {}, however the output tensor is only {}",output_shape_required,  actual_shape));
+}
+
 Tensor moreh_matmul_(
     const Tensor& input_tensor,
     const Tensor& other_tensor,
-    std::optional<std::reference_wrapper<const Tensor>> output_tensor,
+    std::optional<Tensor> output_tensor,
```

Comment:
I have a question.
Is it okay to convert the output_tensor of `std::optional<std::reference_wrapper<const Tensor>>` type to `std::optional<Tensor>` type?